content/slides/code_reviews/_index.md

```diff
@@ -3,24 +3,764 @@ title: "How to review C++ code"
 ---
+# How to review C++ code
 
-# Hello world!
+---
+
+![XKCD](https://imgs.xkcd.com/comics/code_quality.png)
+
+---
+## Why do code reviews?
+
+Code reviews are a great practice. They allow for:
+
+- shared knowledge
+- ensuring code confirms to best practices and the coding standards
+- less smelly code
+- less bugs - cost of a bug grows with the time to find it
+
+---
+
+## What is the most expensive bug that you know of?
+
+[History's worst software bugs](https://www.wired.com/2005/11/historys-worst-software-bugs/)
+
+---
+
+    int connections = GetConnections();
+    if (connections = 2) {
+    }
+
+{{< note >}}
+Could not find the article, but this cost something like 2 mln $ for a UNIX
+distributor.
+{{< /note >}}
+
+---
+## So how to do them?
+
+There are lots of advice on how to conduct code reviews, but it is too general.
+It focuses on:
+
+- how to prepare for the review
+- how to write the review comments
+- enforcing coding standards
+- improving test coverage
+
+---
+## Posting a code review
+
+1. Post a separate review for each meaningful change
+    - unrelated small improvements can get a in-place review
+2. Remove generated files from the review
+    - unless you have changed the code that generates them
+
+---
+## Posting a code review
+
+3. Explain how you have tested the changes
+4. Add any notes that will help the reviewer
+    - link to design document
+    - link to JIRA task
+
+---
+## Posting a code review
+
+> Do review your changes before asking others to review them
+
+---
+## Writing review comments
+
+1. Not personal
+2. Non judgemental
+
+---
+## Review Comments
+
+All of us (and the reviewer and the author of the code) have to keep in mind
+is:
+
+> The code review is about the code produced by a developer at a certain moment
+> in time.
+
+---
+### The reviewer
+
+- You are reviewing the code, not the person.
+
+> "That is will not work when ..." vs "You have missed the case when ..."
+
+> Avoid using *you* in the comments, it is easily replaceable by *we*.
+
+---
+### The author of the code:
+
+- The issues found during the review are issues with the code and
+  implementation. Not with your personality.
+
+> Do better the *next time*.
+
+---
+### Code reviews checklist
+
+
+---
+#### Preration
+
+1. Understand what the code is supposed to do on a high-level and then drill
+   down.
+2. Give yourself enough time - start at 500 LOC / hour and adjust to your speed
+   and knowledge of the area.
+
+---
+#### The review
+
+1. Does the commit message correctly describe the intent of the changes?
+2. Understand what the code actually does. Mark any differences between the
+   intended effects and the actual ones.
+3. Does the code confirm to the coding standards?
+4. Do they follow DRY, KISS?
+5. Are the changes covered by tests?
+6. Are there any seemingly unrelated changes?
+
+---
+#### Extra-care topics
+
+1. Timing issues
+2. Multi-threaded code
+
+Think through all possible scenarios.
+
+TODO: Extend?
```

Comment:
expensive operations in tight loops
blocking operations (e.g. io) in unexpected places